### PR TITLE
Add support for side loading RPC support roles

### PR DESCRIPTION
**Pull request by BjoernT**
This fix side loads the RPC support repository in a Ansible Galaxy style.
pccommon functionatily is originally included in this RPC support repository
and is now automatically pulled down rather than using wget/curl
to pull it down

Connects #1099

### CODE REVIEW

**Comment by git-harry** [comment]
👎 
Please create an issue for this and reference it here.

Why is this written in bash and not created as a playbook?

In principle, I'm not opposed to having such functionality however the action it describes is fairly general and we could end up with multiple of these each for a different group. That makes me wonder if there's a better way.

The commit message is not very professional, please adjust to clearly describe the purpose of the patch.

It's probably worth including documentation explaining the purpose of the code and when to use it.

There also needs to be consideration of if/how/when this would be tested.

**Comment by mancdaz** [comment]
@BjoernT I'm generally OK with this type of stuff going in if it's a convenience script for support and will help you out in a pinch.  However I would want to make the following stipulations:

- All pull requests and commits still need to follow the [contributing guidelines](https://github.com/rcbops/rpc-openstack/blob/master/CONTRIBUTING.md). In the case of this specific PR, the commit message needs updating to add clarity, the PR body needs to reference the issue number, and an issue needs to be created
- This type of convenience script, and any others like it, would not be tested or supported by the dev team, so they would need to live in a separate directory that would be excluded from the testing process. This is actually to give you a lot more freedom and flexibility in the scripts that you drop in here, whilst not burdening the QE or dev team and slowing everyone down - at the end of the day it's a bash wrapper around the ansible code that this repo contains, and you could write as many of these as you like, but we can't commit resource to testing and maintaining all of those.
- Documentation should be included either inline or in some doc note as to the purpose and usage (mostly for the convenience of the support members who would be using it).

An alternative is that these scripts live in a support repo where you manage and maintain their functionality, keeping this repo as a pristine set of tested code. That way you are free to write and maintain these wrapper scripts in the best way that makes sense for you.

As it stands right now based on this and @git-harry comments, I'm 👎 

Thanks

**Comment by mancdaz** [comment]
@BjoernT any thoughts on the above comments? Currently the PR has two -1 reviews so will likely be closed if none of the above are addressed.

**Comment by BjoernT** [comment]
Yeah I have some feedback. Personally I like to use rpc-openstack a module where we can aggregate those type of scripts, I have actually few other scripts and playbooks in mind since this module is always available. Putting those tasks in a separate git repo would be fine too but then we would have to sideload it in rpc-openstack, something I do not know if we can easily extend the ansible role process here.

**Comment by mancdaz** [comment]
Hi @BjoernT I think we're saying it's ok to keep them in here if you like, on the understanding that they will get a once over review when the PR is made, and no subsequent testing after that. If you're ok with this, then you can retarget the PR putting the script inside a 'support-scripts' directory, and add a readme inside there to that effect.

Also, the other comments regarding the commit message etc need to be addressed.  Thanks!

**Comment by BjoernT** [comment]
Updated according feedback

**Comment by mancdaz** [comment]
Thanks @BjoernT 

👍  pending checks

**Comment by mancdaz** [comment]
recheck_ceph

File: support-scripts/rebuild-rabbitmq.sh [hunk 1]
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
+cd /opt/rpc-openstack/openstack-ansible/playbooks
+
+eval $@
```
Comment (hughsaunders) [inline support-scripts/rebuild-rabbitmq.sh]: This is a bad idea. People construct command lines in all sorts of creative ways, and 99% of the time they aren't expecting arguments passed to a script to be executed.

This sort of thing looks inocuous but may catch someone out when they really don't need it (Eg customer down). 

I suggest removing L4 and replacing the test on L5 with ```[[ "$1" != "iknowwhatido" ]]```. L6 will need =1 removing.

**Comment by hughsaunders** [comment]
👎 see line note

**Comment by mancdaz** [comment]
@BjoernT ping

**Comment by Apsu** [comment]
RPC-AIO-Liberty-PR: Fail

**Comment by Apsu** [comment]
RPC-AIO-Ceph-Liberty-PR: Fail

**Comment by BjoernT** [comment]
recheck

**Comment by Apsu** [comment]
RPC-AIO-Ceph-Liberty-PR: Fail

**Comment by Apsu** [comment]
RPC-AIO-Liberty-PR: Fail

**Comment by BjoernT** [comment]
recheck_all

File: rpcd/playbooks/rpc-support.yml [hunk 2]
```diff
@@ -18,3 +18,12 @@
   user: root
   roles:
     - { role: "rpc_support", tags: [ "rpc-support" ] }
+
+- name: Drop pccommon script
```
Comment (mancdaz) [inline rpcd/playbooks/rpc-support.yml]: presumably at some point this role will do more than just install pccommon from it's main.yml? I'd suggest a more generic name for this play (e.g. 'install support scripts'), and the task inside the role should be named for what it's doing (i.e. 'install pccommon script' or whatever).

**Comment by mancdaz** [comment]
@BjoernT the role is failing ansible linting. In order to re-run the travis jobs after you fix it, you'll need to click 'details' on the travis job link above and click 'rebuild job' inside there.  recheck_* doesn't re-run travis linting jobs.

Also, if the jenkins build jobs fail, it's worth following the links and noting what they failed on here in the PR so that we can be sure we're not rechecking blindly on an actual issue with the PR

**Comment by BjoernT** [comment]
@mancdaz Yes I fixed few of them yesterday. I'll have a look

**Comment by Apsu** [comment]
RPC-AIO-Ceph-Liberty-PR: Fail

**Comment by Apsu** [comment]
RPC-AIO-Liberty-PR: Fail

**Comment by BjoernT** [comment]
recheck_all

**Comment by Apsu** [comment]
RPC-AIO-Liberty-PR: Fail

**Comment by Apsu** [comment]
RPC-AIO-Ceph-Liberty-PR: Fail

**Comment by Apsu** [comment]
RPC-AIO-Liberty-PR: Fail

**Comment by Apsu** [comment]
RPC-AIO-Ceph-Liberty-PR: Fail

**Comment by BjoernT** [comment]
recheck_all

**Comment by Apsu** [comment]
RPC-AIO-Ceph-Liberty-PR: Fail

**Comment by Apsu** [comment]
RPC-AIO-Liberty-PR: Fail

**Comment by hughsaunders** [comment]
Last jobs failed on portnotfoundclient which should now be fixed. recheck_all

**Comment by BjoernT** [comment]
recheck_all

**Comment by BjoernT** [comment]
@prometheanfire @d34dh0r53 Please have a look if we can merge it, I like to proceed adding more tasks once it merges. Currently I only offloaded pccommon out of the rpc-support role but plan to retire it entirely soon

**Comment by prometheanfire** [comment]
I'm generally OK with this, but don't know what it should target (if all active branches are good or not), I'll raise that in the parent issue.

:+1:

**Comment by git-harry** [comment]
@BjoernT @prometheanfire this work is not currently approved and so the dev team should not be investing time in it. Please raise this at the triage meeting if you wish to see priorities adjusted.

/cc @b3rnard0

**Comment by BjoernT** [comment]
recheck_all

**Comment by prometheanfire** [comment]
recheck_upgrade

**Comment by prometheanfire** [comment]
recheck_upgrade

**Comment by prometheanfire** [comment]
recheck_upgrade

**Comment by prometheanfire** [comment]
recheck_upgrade

**Comment by git-harry** [comment]
👎 This pull request seems very different to what had previously been discussed and now looks to be mostly a refactor of the pccommon install code with the ability to install a RAID controller utility.

- What does side-loading refer to in this context?
-  What benefit does this role provide?
- Why was arcconf never added to rpc_support instead of creating a separate role?

**Comment by BjoernT** [comment]
Hi @git-harry as always said this role is there for Support to start automating stuff and contributing to rpc-o has traditionally be a difficult task. Splitting the ops from dev allows us ops people to contribute in a different speed and later merge with the product via side loading, ansible galaxy style. I'm pretty sure you know the reasons why we want to go galaxy style route.

**Comment by BjoernT** [comment]
recheck_upgrade

**Comment by git-harry** [comment]
@BjoernT 

What does side-loading refer to in this context? It is not a term that I have ever seen used in the context of Ansible and so I would like to be clear as to what you are trying to say.

If I understand you right, you're suggesting that you have found contributing to RPCO a difficult task and so this is an attempt to workaround it. I think it would have been better to attempt to explore that more first. What is the process by which you think this role should be updated in RPCO? How often should it be updated? Who is responsible when the testing fails due to bugs/incompatibility?

What processes are in place to manage rcbops_openstack-ops? What testing do you have set up? How are code reviews managed? How is work (bugs/features) for the role tracked and prioritised?

The original discussion related to the use of 'helper' scripts, the role is not copying any scripts, it is installing something already performed by rpc_support and something else that has never been raised as a requirement. How does rcbops_openstack-ops compare with rpc_support? How would you define their purposes?

**Comment by sigmavirus24** [comment]
@git-harry I understand this also as a way to side-step contribution guidelines to rpc-openstack by having a place that only support controls.

Looking at the repository, it appears pull requests are not used (https://github.com/rsoprivatecloud/openstack-ops/pulls?utf8=✓&q=is:pr%20) as there are none. Likewise, issues are not used either (https://github.com/rsoprivatecloud/openstack-ops/issues?utf8=✓&q=is:issue). The role does not contain any testing except for running ansible-lint (https://github.com/rsoprivatecloud/openstack-ops/blob/9b39e4d9b6f9be60dd353e2b4d41fd0c7ea7749a/.travis.yml).

I'm not convinced that this needs to live in our list of galaxy-role-requirements and the role (as you already noted) does far more than install pccommon tooling.

All of that said, I'm not opposed to support maintaining their own automation as they see fit, but I'm not convinced it needs to be hooked into this repository via "side-loading". Support can already install this role and run their own playbook to perform these actions. I full well understand that this solution is inconvenient, but I'm not convinced this actually belongs here as it is.

If there are issues with pccommon, they've not been brought to our attention https://github.com/rcbops/rpc-openstack/issues?q=is:issue+pccommon+is:open, and all of the closed issues are from almost a year (or more) ago and seem to have been resolved (which I infer only from the fact that they are and have stayed closed).

It seems it would have been more efficient to raise these issues and feature requests and that it's still possible to do so. We can integrate this work directly into rpc-o with proper review and testing.

**Comment by BjoernT** [comment]
@sigmavirus24 @git-harry  Thanks for the feedback but as I said it's not about pcommon at all more about the automation I have in mind to fix our SOP issues. pccommon should not be a part of rpc-o in the first place since it's pure support oriented. In the end, using ansible galaxy roles for this is a good way to combine my needs. If we don't use that, what else?. Because the alternatives would be not to use the rpc-o scripts to deploy rpc-o. De facto stop using this repo and I need to write support scripts which I'm fine with that but that would mean we in support would reduce to just take your roles and fork rpc-o into a support repo. I don't really want to fork rpc-o since I see support as primary user for this product, if that is not the case then OK I will fork it. But let's be honest here, the barrier to contribute to rpc-o is so high that if you want to offer merging rpc-o and our openstack-ops module then we have to talk candidly about the contribution problem and the ever changing guide lines, creating barriers. This PR is open for 6 month and I can't afford to progress that slow in our business. I'm adding @claco to get his view on what rpc-o should be, perhaps it is not really clear to me/us.

**Comment by sigmavirus24** [comment]
> I don't really want to fork rpc-o since I see support as primary user for this product

To the best of my knowledge, @BjoernT, support **is** the primary user of rpc-o. That's why there is literally a role in rpc-o called "rpc_support". That's where pccommon's installation lived. That's where we should be adding these things to.

> then we have to talk candidly about the contribution problem and the ever changing guide lines, creating barriers.

Yes, let's talk about that. We've tried everything (including meetings, emails, and documentation in this very repository) to announce, explain, and train the guidelines. What was missed? How can we communicate these things to you so that you absorb them and understand them and can work with them?

**Comment by alextricity25** [comment]
This was discussed in today's triage with support(Steve Ribble, @coldcolors, and @nategraf) . While we do understand the need for support to drop in the their own tooling, we want to get a better understanding about the process and what we can improve upon before this PR merges. This will involve getting the proper people engaged in conversations as to how we can make it easier on support in respect to tooling and issues that need to be addressed.
In the meantime, I'm going to mark this as blocked until we have those conversations.
cc. @claco @BjoernT @b3rnard0

**Comment by b3rnard0** [comment]
@jamiramontez Also need you to weigh in on this issue.

**Comment by BjoernT** [comment]
recheck_upgrade

**Comment by BjoernT** [comment]
@alextricity25 Based on your feedback do you have questions and which one do you have ?

**Comment by alextricity25** [comment]
Hi @BjoernT. Thanks for the quick reply.
No questions, just a few concerns that we wish to get some more insight on before we merge this. Ultimately, it comes down to maintaining and having visibility into what's being merged into the new role . Right now, these changes [include the new role as part of the rpc-support.yml playbook](https://github.com/rcbops/rpc-openstack/pull/1035/files#diff-2f10d4a70098aa4492fb851418dcc0a5R26). That means that the tasks in this new role will be run where ever rpc-support.yml is run. If something merges in the ``openstack-ops`` role and it's not tested, then it poses the risk of breaking our gate jobs/deployment. This is mainly where most of our concern is coming from.
Even though this commit has changed a bit since it's inception, the comment @mancdaz made [here](https://github.com/rcbops/rpc-openstack/pull/1035#issuecomment-219422595) still needs to addressed at a higher level. @sigmavirus24 also brought up a few good points [here](https://github.com/rcbops/rpc-openstack/pull/1035#issuecomment-259710721).

I strongly believe that once support can convince us those items listed in Darren's comment will be taken care of, then this PR won't have a problem merging. We'll go ahead and set something up where we can have those discussions.
cc @claco @b3rnard0 @coldcolors

**Comment by sigmavirus24** [comment]
One thing I failed to note in my original comment is the following:

Presently, we don't have very much testing for our `rpc_support` role in this project. You could say that we test it by at least ensuring it runs without failing but that's a bare minimum of testing. The role you're including, by comparison, is worse off than that. There's configuration for running ansible-lint via Travis-CI but other than that, there's no automated testing of the role.

@BjoernT do you all plan to add any automated testing to that role?

**Comment by BjoernT** [comment]
@alextricity25 This role is supposed to eliminate rpc_support over time and more so to add support SOPs which we can develop in a different speed than rpc-o but ultimately synchronizing with the ansible role requirements and frozen via SHA. If you are still concerned about that this role break this gating then I will close this PR and move on and consume RPC-O differently.

Also it is planned to have a functional testing for the default tasks. All auxiliary tasks, which will follow can not be part of this RPC-O gating because they would interfere like rebuilding rabbit container, galera backup-restore automation etc. 

In regards to current rpc_support testing, yes when the playbooks finish that is enough testing for me.

Lastly I don't currently see what else is needed related to @mancdaz  comment beyond the newly added release notes. At some point I would much rather prefer to know the scope, before investing too much time and then realizing it will never merge.

**Comment by alextricity25** [comment]
Based on the meeting we had today, this role will solely contain administrative tasks that will help support drop in their tools. The understanding is that support will maintain the the role[1] and test coverage will not be provided by Dev nor QE. 
Additionally, we will continue a discussion offline as to how we can make our contributing process more fluid to outside contributors. @BjoernT please send me any items that you think we can improve upon and we will take a closer look.

As of now, I am all for merging this PR and making support's life easier to drop any administrative tasks they wish to perform on an environment. Again, this is with the understanding that support will be responsible for the quality and the maintaining of the external role. :+1:
If anyone from the meeting has anything to add, please feel free to. Thanks!
cc. @claco @b3rnard0 @mancdaz @weezer @Rydor @sigmavirus24 @go2jumping 

[1]https://github.com/rsoprivatecloud/openstack-ops

**Comment by git-harry** [comment]
@alextricity25 so what is the purpose of rpc_support and how does it compare with this new role? rpc_support is there to include things that support want - should that now be offloaded to support with the expectation they take any action required in this new role?

As an example of the confusion of having two roles, rpc_support installs some RAID tools, this new role installs a RAID tool (arcconf). It appears arcconf could be installed by overriding an existing var in rpc_support, so there is duplication. Leaving this as is makes it confusing as to where the responsibility lies. In addition, I can find no request to be able to install it by default. Where was the deficiency in the contributing process?

If this change goes ahead, what open issues can be transferred to https://github.com/rsoprivatecloud/openstack-ops?

@BjoernT please provide details of the shortcomings you perceive in the process. I do not recollect you raising any concerns over the last 6 months, what has changed?

**Comment by BjoernT** [comment]
@git-harry I'm not sure if you were part of the meeting this morning, rpc_support role will be moved into the openstack-ops repo with the next update. Only the playbooks remain to call roles

**Comment by prometheanfire** [comment]
so where do we stand here?

**Comment by d34dh0r53** [comment]
I don't know where we stand on this politically but the code looks good to me so 👍

**Comment by alextricity25** [comment]
> @alextricity25 so what is the purpose of rpc_support and how does it compare with this new role? rpc_support is there to include things that support want - should that now be offloaded to support with the expectation they take any action required in this new role?

Correct. ``rpc_support`` will be moved over to the new role being pulled in via role-requirements. There is a mutual understanding that support will own that code, and is responsible for the quality and testing of the code. No test coverage, bug fixes, or features to this external repo will be provided by the Dev team. The external repo should only contain administration tasks that support deems necessary, and should not in anyway effect the functionality nor architecture of the RPCO product.

Given that this has been review by two folks, I will go ahead and merge this in.

**Pull request merged**